File: app/src/main/java/it/amonshore/comikkua/data/web/AvailableComics.java
package it.amonshore.comikkua.data.web;

import android.text.TextUtils;

import com.google.gson.annotations.SerializedName;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.room.Entity;
import androidx.room.Index;
import androidx.room.PrimaryKey;

/**
 * Rappresenta un comics disponibile per essere monitorato:
 * cioè le release possono essere aggiornate in automatico.
 * Una volta monitorato viene creato un record in tComcis con sourceId
 * valorizzato a tAvailableComics.id.
 */
@Entity(tableName = "tAvailableComics",
        indices = {@Index(value = "sourceId", unique = true),
                @Index({"name", "publisher", "version"})})
public class AvailableComics {

    public final static long NO_COMICS_ID = -1;

    /**
     * Id comics interno.
     * Room supporta solo id di tipo long.
     */
    @PrimaryKey(autoGenerate = true)
    public long id;

    /**
     * Id del comics così come indentificato in rete.
     */
    @SerializedName("id")
    public String sourceId;

    /**
     * Nome del cmomics.
     */
    @SerializedName("name")
    public String name;

    /**
     * Nome del cmomics usato per le ricerche.
     */
    @SerializedName("searchableName")
    public String searchableName;

    /**
     * Editore.
     */
    @SerializedName("publisher")
    public String publisher;

    /**
     * Versione del comics, cioè il numero di ristampa.
     * 0=nessuna ristampa, 1=prima ristampa, etc.
     */
    @SerializedName("version")
    public int version;

    public AvailableComics withSourceId(String sourceId) {
        this.sourceId = sourceId;
        return this;
    }

    @NonNull
    public String getInitial() {
        if (TextUtils.isEmpty(name)) {
            return "";
        } else {
            return name.substring(0, 1);
        }
    }

    @Override
    public boolean equals(@Nullable Object obj) {
        if (obj == this) return true;

        if (obj instanceof AvailableComics) {
            final AvailableComics other = (AvailableComics) obj;
            return other.sourceId.equals(this.sourceId);
        } else {
            return false;
        }
    }
}

File: app/src/main/java/it/amonshore/comikkua/data/web/AvailableComics.kt
package it.amonshore.comikkua.data.web

import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.Index
import androidx.room.PrimaryKey
import com.google.gson.annotations.SerializedName

/**
 * Rappresenta un comics disponibile per essere monitorato:
 * cioè le release possono essere aggiornate in automatico.
 * Una volta monitorato viene creato un record in tComcis con sourceId
 * valorizzato a tAvailableComics.id.
 */
@Entity(
    tableName = "tAvailableComics",
    indices = [Index("sourceId", unique = true), Index("name", "publisher", "version")]
)
data class AvailableComics(
    @SerializedName("ref_id") val sourceId: String,
    @SerializedName("title") val name: String,
    @SerializedName("editor") val publisher: String,
    @SerializedName("reissue") val version: Int,
) {
    /**
     * Id comics interno.
     * Room supporta solo id di tipo long.
     */
    @JvmField
    @PrimaryKey(autoGenerate = true)
    var id: Long = 0

    val initial: String
        get() = if (name.isEmpty()) "" else name.substring(0, 1)

    @delegate:Ignore
    val searchableName: String by lazy {
        "${name.uppercase()} ${publisher.uppercase()}"
    }

    override fun equals(other: Any?): Boolean {
        if (other === this) return true
        return if (other is AvailableComics) {
            other.sourceId == sourceId
        } else {
            false
        }
    }

    override fun hashCode(): Int {
        return sourceId.hashCode()
    }
}